subsys/net/lib/sockets/sockets_can.c

```diff
@@ -0,0 +1,353 @@
+			net_pkt_set_eof(last_pkt, true);
+			NET_DBG("Set EOF flag on pkt %p", ctx);
+		}
+
+		return;
+	}
+
+	/* Normal packet */
+	net_pkt_set_eof(pkt, false);
+
+	k_fifo_put(&ctx->recv_q, pkt);
+}
+
+static int zcan_bind_ctx(struct net_context *ctx, const struct sockaddr *addr,
+			 socklen_t addrlen)
+{
+	struct sockaddr_can *can_addr = (struct sockaddr_can *)addr;
+	struct net_if *iface;
+	int ret;
+
+	if (addrlen != sizeof(struct sockaddr_can)) {
+		return -EINVAL;
+	}
+
+	iface = net_if_get_by_index(can_addr->can_ifindex);
+	if (!iface) {
+		return -ENOENT;
+	}
+
+	net_context_set_iface(ctx, iface);
+
+	ret = net_context_bind(ctx, addr, addrlen);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	/* For CAN socket, we expect to receive packets after call to bind().
+	 */
+	ret = net_context_recv(ctx, zcan_received_cb, K_NO_WAIT,
+			       ctx->user_data);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	return 0;
+}
+
+ssize_t zcan_sendto_ctx(struct net_context *ctx, const void *buf, size_t len,
+			int flags, const struct sockaddr *dest_addr,
+			socklen_t addrlen)
+{
+	struct sockaddr_can can_addr;
+	s32_t timeout = K_FOREVER;
+	int ret;
+
+	/* Setting destination address does not probably make sense here so
+	 * ignore it. You need to use bind() to set the CAN interface.
+	 */
+	if (dest_addr) {
+		NET_DBG("CAN destination address ignored");
+	}
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	if (addrlen == 0) {
+		addrlen = sizeof(struct sockaddr_can);
+	}
+
+	if (dest_addr == NULL) {
+		memset(&can_addr, 0, sizeof(can_addr));
+
+		can_addr.can_ifindex = -1;
+		can_addr.can_family = AF_CAN;
+
+		dest_addr = (struct sockaddr *)&can_addr;
+	}
+
+	ret = net_context_sendto_new(ctx, buf, len, dest_addr, addrlen, NULL,
+				     timeout, NULL, ctx->user_data);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	return len;
+}
+
+static ssize_t zcan_recvfrom_ctx(struct net_context *ctx, void *buf,
+				 size_t max_len, int flags,
+				 struct sockaddr *src_addr,
+				 socklen_t *addrlen)
+{
+	size_t recv_len = 0;
+	s32_t timeout = K_FOREVER;
+	struct net_pkt *pkt;
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	if (flags & ZSOCK_MSG_PEEK) {
+		int ret;
+
+		ret = _k_fifo_wait_non_empty(&ctx->recv_q, timeout);
+		/* EAGAIN when timeout expired, EINTR when cancelled */
+		if (ret && ret != -EAGAIN && ret != -EINTR) {
+			errno = -ret;
+			return -1;
+		}
+
+		pkt = k_fifo_peek_head(&ctx->recv_q);
+	} else {
+		pkt = k_fifo_get(&ctx->recv_q, timeout);
+	}
+
+	if (!pkt) {
+		errno = EAGAIN;
+		return -1;
+	}
+
+	/* We do not handle any headers here, just pass the whole packet to
+	 * the caller.
+	 */
+	recv_len = net_pkt_get_len(pkt);
+	if (recv_len > max_len) {
+		recv_len = max_len;
+	}
+
+	if (net_pkt_read_new(pkt, buf, recv_len)) {
+		errno = EIO;
+		return -1;
+	}
+
+	if (!(flags & ZSOCK_MSG_PEEK)) {
+		net_pkt_unref(pkt);
+	}
```

Comment:
else { net_pkt_cursor_init(pkt); }  so another msg peek on the same pkt will work